File: lib/services/attendance_service.dart
import 'package:flutter/material.dart';
import '../data/database/app_database.dart';
import '../data/repositories/repositories.dart';
import 'wifi_service.dart';
import '../../core/utils/date_utils.dart';

/// Result of attendance check
enum AttendanceCheckResult {
  success, // Marked successfully
  alreadyMarked, // Already marked today
  wifiMismatch, // Not connected to office WiFi
  noOfficeConfig, // Office not configured
  nonWorkingDay, // Today is not a working day
  error, // Other error
}

/// Service for handling automated Wi-Fi attendance marking logic
class AttendanceService {
  final WifiService _wifiService;
  
  AttendanceService({
    WifiService? wifiService,
  })  : _wifiService = wifiService ?? WifiService();
  
  /// Checks and automatically marks attendance if connected to office Wi-Fi network
  Future<AttendanceCheckResult> checkAndMarkAttendance({
    required OfficeConfig officeConfig,
    required AttendanceRepository attendanceRepo,
    required Function() onAttendanceMarked,
  }) async {
    try {
      // Check if already marked today
      final today = getCurrentDateYyyyMmDd();
      final existingRecord = await attendanceRepo.getByDate(today);
      if (existingRecord != null) {
        return AttendanceCheckResult.alreadyMarked;
      }
      
      // Check if today is a working day
      final now = DateTime.now();
      if (!isWorkingDay(now, officeConfig.workingDaysMask)) {
        return AttendanceCheckResult.nonWorkingDay;
      }
      
      // Check WiFi SSID
      final isOnOfficeWifi = await _wifiService.isConnectedToSSID(officeConfig.ssid);
      if (!isOnOfficeWifi) {
        return AttendanceCheckResult.wifiMismatch;
      }
      
      // Attendance is automatically marked as Present whenever Wi-Fi is connected
      const status = AttendanceStatus.present;

      
      // Automatically mark attendance
      await attendanceRepo.insertRecord(
        dateYyyyMmDd: today,
        status: status,
        markedAt: DateTime.now(),
        ssidSnapshot: await _wifiService.getWifiSSID(),
        distanceMeters: null,
      );
      
      // Callback for notification
      onAttendanceMarked();
      
      return AttendanceCheckResult.success;
    } catch (e) {
      print('Error checking automated attendance: $e');
      return AttendanceCheckResult.error;
    }
  }
  
  /// Gets current WiFi status
  Future<Map<String, dynamic>> getWifiStatus() async {
    try {
      final ssid = await _wifiService.getWifiSSID();
      final isConnected = ssid != null;
      
      return {
        'isConnected': isConnected,
        'ssid': ssid,
      };
    } catch (e) {
      print('Error getting WiFi status: $e');
      return {'isConnected': false, 'ssid': null};
    }
  }
}

File: app/src/main/kotlin/com/urunkarpm/pingpin/service/AttendanceService.kt
package com.urunkarpm.pingpin.service

import android.content.Context
import android.util.Log
import com.urunkarpm.pingpin.data.local.entity.OfficeConfigEntity
import com.urunkarpm.pingpin.data.repository.AttendanceRepository
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.Locale

enum class AttendanceCheckResult {
    SUCCESS,
    ALREADY_MARKED,
    WIFI_MISMATCH,
    NO_OFFICE_CONFIG,
    NON_WORKING_DAY,
    ERROR
}

object WorkingDays {
    const val MONDAY = 1
    const val TUESDAY = 2
    const val WEDNESDAY = 4
    const val THURSDAY = 8
    const val FRIDAY = 16
    const val SATURDAY = 32
    const val SUNDAY = 64

    const val DEFAULT_WEEKDAYS = MONDAY or TUESDAY or WEDNESDAY or THURSDAY or FRIDAY

    fun isWorkingDay(calendar: Calendar, workingDaysMask: Int): Boolean {
        // Calendar.DAY_OF_WEEK: Sunday=1, Monday=2, Tuesday=3, Wednesday=4, Thursday=5, Friday=6, Saturday=7
        val dayShift = when (calendar.get(Calendar.DAY_OF_WEEK)) {
            Calendar.MONDAY -> 0
            Calendar.TUESDAY -> 1
            Calendar.WEDNESDAY -> 2
            Calendar.THURSDAY -> 3
            Calendar.FRIDAY -> 4
            Calendar.SATURDAY -> 5
            Calendar.SUNDAY -> 6
            else -> 0
        }
        return (workingDaysMask and (1 shl dayShift)) != 0
    }

    fun isWfoDay(calendar: Calendar, wfoDaysMask: Int): Boolean {
        val dayShift = when (calendar.get(Calendar.DAY_OF_WEEK)) {
            Calendar.MONDAY -> 0
            Calendar.TUESDAY -> 1
            Calendar.WEDNESDAY -> 2
            Calendar.THURSDAY -> 3
            Calendar.FRIDAY -> 4
            Calendar.SATURDAY -> 5
            Calendar.SUNDAY -> 6
            else -> 0
        }
        return (wfoDaysMask and (1 shl dayShift)) != 0
    }
}

open class AttendanceService(
    private val context: Context,
    private val wifiService: WifiService = WifiService(context)
) {
    companion object {
        private const val TAG = "AttendanceService"

        fun getCurrentDateYyyyMmDd(): String {
            val sdf = SimpleDateFormat("yyyy-MM-dd", Locale.US)
            return sdf.format(Date())
        }
    }

    open suspend fun checkAndMarkAttendance(
        officeConfig: OfficeConfigEntity,
        attendanceRepo: AttendanceRepository,
        onAttendanceMarked: suspend () -> Unit
    ): AttendanceCheckResult {
        return try {
            val today = getCurrentDateYyyyMmDd()

            // 1. Check if already marked today
            val existing = attendanceRepo.getByDate(today)
            if (existing != null) {
                return AttendanceCheckResult.ALREADY_MARKED
            }

            // 2. Check working days
            val cal = Calendar.getInstance()
            if (!WorkingDays.isWorkingDay(cal, officeConfig.workingDaysMask)) {
                return AttendanceCheckResult.NON_WORKING_DAY
            }

            // 3. Check Wi-Fi match
            val isOnWifi = wifiService.isConnectedToSSID(officeConfig.ssid)
            if (!isOnWifi) {
                return AttendanceCheckResult.WIFI_MISMATCH
            }

            // 4. Record attendance as "present"
            val currentSsid = wifiService.getWifiSSID()
            attendanceRepo.insertRecord(
                dateYyyyMmDd = today,
                status = "present",
                markedAt = System.currentTimeMillis(),
                ssidSnapshot = currentSsid,
                distanceMeters = null
            )

            // Trigger notification callback
            onAttendanceMarked()

            AttendanceCheckResult.SUCCESS
        } catch (e: Exception) {
            Log.e(TAG, "Error checking attendance", e)
            AttendanceCheckResult.ERROR
        }
    }
}
